### ISSUE TITLE
Submitted succeeded flag

### ISSUE THREAD

**Issue by morgante** [erikras/redux-form#527]
It would be helpful to have a prop which is set to true once a form has been submitted successfully. Essentially the inverse of #276 (except both start off false).

This would be helpful for confirming to the user that the form has been successfully saved.

**Comment by erikras** [comment]
You can know that if your submit promise resolves. You don't really want a permanent `true` boolean from then on, because you don't want the `Saved!` message present forever and ever. That's why that state management is left to the library user.

You really see some value in having a "one save went through in the history of this form" flag?

**Comment by morgante** [comment]
> You can know that if your submit promise resolves.

How exactly?

> You don't really want a permanent true boolean from then on, because you don't want the Saved! message present forever and ever.

With the new behavior of destroying on unmount, it wouldn't be forever. It's just until the user leaves the form or I otherwise reset it.

I definitely see value in knowing whether this form successfully saved since it's last initialization/reset. It seems like the natural corollary to the `submitFailed` flag.

i realize it would be possible to do this with a custom reducer, but having to manage that state (including tracking initializations) seems excessively complicated for the common use case of a success confirmation.

**Comment by drouillard** [comment]
FWIW I would leave this as is. The more common use case is for the confirmation message to 'flash' for a certain amount of time. The message, how long it should flash, and the styling are all likely custom per user. Also was not hard to add (just did it)

**Comment by erikras** [comment]
> How exactly?

```js
<button onClick={() =>
  this.props.handleSubmit()
    .then(() => {
      this.setState({saved: true});
      setTimeout(() => {
        this.setState({saved: false});
      }, 5000); // show message for 5 seconds
   })
}>Save</button>
{this.state.saved && <div>Saved!</div>}
```

**Comment by morgante** [comment]
Okay, I didn't realize that handleSubmit returned a promise. I still don't understand why `submitFailed` is useful but `submitSucceeded` isn't.

**Comment by erikras** [comment]
Mainly because the submit can fail through various paths...

1. Sync validation invalid
2. Async validation invalid
3. Submit validation invalid
4. Some other server side error

...and it can only succeed through one. And most often a successful form submit will also involve some other use-case-specific state change, like navigating back to a list, or changing all the inputs to read-only, or some dialog that is shown for a short period of time. But an error just needs to stop and say that something went wrong, regardless of what the app has planned for post-successful-submit.

**Comment by spudly** [comment]
For anyone looking for a solution to this problem, here's how I implemented a solution in my project:

1. Add a reducer for submitSucceeded:

```js
import {reducer as formReducer, actionTypes} from 'redux-form';
import {createStore, combineReducers} from 'redux';

const reducer = combineReducers({
  form: formReducer,
  submitSucceeded: (submitSucceeded = false, action) => {
    switch (action.type) {
      case actionTypes.TOUCH: return false;
      case actionTypes.START_SUBMIT: return false;
      case actionTypes.STOP_SUBMIT: return action.errors ? false : true;
      default: return submitSucceeded;
    }
  },
});
```

2. Add (or modify) a `mapStateToProps` function that selects the new `submitSucceeded` flag from the state:

```js
const mapStateToProps = state => {
  return {
    submitSucceeded: state.submitSucceeded,
  };
};

export default reduxForm(formConfig, mapStateToProps)(MyFormComponent);
```

Now you'll have an additional prop in your component called `submitSucceeded`.

I still think that this flag should be added to `redux-form`, but this code works around the issue by creating our own flag.

**Comment by morgante** [comment]
@spudly How do you deal with destroying the submitted property when you navigate to a new form?

I still think this should be a built-in flag.

**Comment by spudly** [comment]
I didn't have that problem, but o would just create another action creator
that gets dispatched when the component unmounts. Then set the property to
false on your reducer when it is dispatched.

**Comment by LiranBri** [comment]
@erikras following your suggestion with registering the handleSubmit's returned promise, what if i want to pass handleSubmit my own callback? i can't pass both "data" and my callback to handleSubmit

**Comment by sergeyzwezdin** [comment]
I would agree with @morgante - what the problem to add the flag?
Once submit is done, it's switched to true. After that once any field is touched, it's cleared to false.
Anyone who doesn't like it can just ignore this flag.

**Comment by morgante** [comment]
@ooflorent Why was this closed? It seems like a reasonable request still.

**Comment by ooflorent** [comment]
As Erik said:

> You can know that if your submit promise resolves. You don't really want a permanent `true` boolean from then on, because you don't want the _Saved!_ message present forever and ever. That's why that state management is left to the library user.

Moreover, @spudly demonstrated how do it using a custom reducer.

Another issue (#853) was later opened asking for this flag. It should have been closed in favor of this one but it catches my eye first! Keeping both opened is not necessary.

`v5.1.0` adds a new config parameter to kinda achieve this: `onSubmitSuccess`. It could also be used to set `submitSucceed` flag.

I closed this issue for all the reason above.

---

However, I would be really happy to review a PR adding this flag! If you need help or assist, feel free to ask!

**Comment by morgante** [comment]
Okay, so this is now a duplicate of #853.

**Comment by ooflorent** [comment]
Yes, sorry, if I wasn't clear about it.

**Comment by vladhd93** [comment]
@erikras 
```<button onClick={(data) =>
  this.props.handleSubmit(data)
    .then(() => {
      this.setState({saved: true});
      setTimeout(() => {
        this.setState({saved: false});
      }, 5000); // show message for 5 seconds
   })
}>Save</button>
{this.state.saved && <div>Saved!</div>}
```  
but this example not working, handleSubmit don't return promise
```Uncaught TypeError: Cannot read property 'then' of undefined```

**Comment by xinghengwang** [comment]
@vladhd93  what does your implementation of handleSubmit look like?  Suppose you implemented your own handleSubmit (as it is required), it needs to return a promise for the above code snippet to work.

**Comment by tranvansang** [comment]
Now `redux-form` supports `submitSucceeded` flag.

```javascript
MyComponent = ({submitSucceeded}) => <div>
// other elements
{submitSucceeded && <div className="alert alert-success" role="alert">Saved</div>}
</div>
```
